In [1]:
from __future__ import print_function
import numpy as np
np.random.seed(1337)  # for reproducibility

from keras.preprocessing import sequence
from keras.models import Sequential
from keras.layers import Dense, Dropout, Activation, Lambda
from keras.layers import Embedding
from keras.layers import Convolution1D,MaxPooling1D, Flatten
from keras.datasets import imdb
from keras import backend as K
from sklearn.model_selection import train_test_split
import pandas as pd
from keras.utils.np_utils import to_categorical
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import Normalizer
from keras.models import Sequential
from keras.layers import Convolution1D, Dense, Dropout, Flatten, MaxPooling1D
from keras.utils import np_utils
import numpy as np
import h5py
from keras import callbacks
from keras.layers import LSTM, GRU, SimpleRNN, Bidirectional
from keras.callbacks import CSVLogger
from keras.callbacks import ModelCheckpoint, EarlyStopping, ReduceLROnPlateau, CSVLogger

Using TensorFlow backend.


In [2]:
traindata = pd.read_csv('/home/ritika/NetworkIntrusionDetection/UNSW-NB15 - CSV Files/a part of training and testing set/UNSW_NB15_training-set.csv')
testdata = pd.read_csv('/home/ritika/NetworkIntrusionDetection/UNSW-NB15 - CSV Files/a part of training and testing set/UNSW_NB15_testing-set.csv')

In [3]:
#traindata=traindata.drop('id',axis=1)
#testdata=testdata.drop('id',axis=1)

from sklearn.preprocessing import LabelEncoder
lb_make = LabelEncoder()

traindata['proto'] = lb_make.fit_transform(traindata['proto'])
traindata['service'] = lb_make.fit_transform(traindata['service'])
traindata['state'] = lb_make.fit_transform(traindata['state'])
traindata['attack_cat'] = lb_make.fit_transform(traindata['attack_cat'])

testdata['proto'] = lb_make.fit_transform(testdata['proto'])
testdata['service'] = lb_make.fit_transform(testdata['service'])
testdata['state'] = lb_make.fit_transform(testdata['state'])
testdata['attack_cat'] = lb_make.fit_transform(testdata['attack_cat'])

X = traindata.iloc[:,1:42]
Y = traindata.iloc[:,-2]
x = testdata.iloc[:,1:42]
y = testdata.iloc[:,-2]

In [4]:
Arr = X.values
arr = x.values

In [5]:
import librosa
Arr_Delta = librosa.feature.delta(Arr)
arr_delta = librosa.feature.delta(arr)

Delta1 = pd.DataFrame(Arr_Delta)
delta1 = pd.DataFrame(arr_delta)

In [6]:
XDelta1 = pd.concat([X, Delta1], axis=1)
xdelta1 = pd.concat([x, delta1], axis=1)

scaler = Normalizer().fit(XDelta1)
XDelta1 = scaler.transform(XDelta1)
scaler = Normalizer().fit(xdelta1)
xdelta1 = scaler.transform(xdelta1)

Y = np.array(Y)
y = np.array(y)

XDelta1 = np.reshape(XDelta1, (XDelta1.shape[0],XDelta1.shape[1],1))
xdelta1 = np.reshape(xdelta1, (xdelta1.shape[0],xdelta1.shape[1],1))

lstm_output_size = 70

In [7]:
Y = to_categorical(Y)
y = to_categorical(y)

In [15]:
lstm_output_size = 70

cnn = Sequential()
cnn.add(Convolution1D(64, 3, border_mode="same",activation="relu",input_shape=(82, 1)))
cnn.add(MaxPooling1D(pool_size=(2)))
cnn.add(Bidirectional(LSTM(lstm_output_size)))
cnn.add(Dropout(0.1))
cnn.add(Dense(10, activation="softmax"))

cnn.compile(loss="categorical_crossentropy", optimizer="adam",metrics=['accuracy'])

cnn.fit(XDelta1, Y, epochs=200, validation_data=(xdelta1, y))

/home/ritika/.virtualenvs/project/lib/python3.5/site-packages/ipykernel_launcher.py:4: UserWarning: Update your `Conv1D` call to the Keras 2 API: `Conv1D(64, 3, input_shape=(82, 1), activation="relu", padding="same")`
  after removing the cwd from sys.path.


Train on 175341 samples, validate on 82332 samples
Epoch 1/200
175341/175341 [==============================] - 270s 2ms/step - loss: 1.3819 - accuracy: 0.5282 - val_loss: 1.2940 - val_accuracy: 0.5931
Epoch 2/200
175341/175341 [==============================] - 257s 1ms/step - loss: 1.3362 - accuracy: 0.5432 - val_loss: 1.2930 - val_accuracy: 0.5942
Epoch 3/200
175341/175341 [==============================] - 294s 2ms/step - loss: 1.3253 - accuracy: 0.5474 - val_loss: 1.2779 - val_accuracy: 0.5870
Epoch 4/200
175341/175341 [==============================] - 326s 2ms/step - loss: 1.3195 - accuracy: 0.5492 - val_loss: 1.2795 - val_accuracy: 0.5802
Epoch 5/200
175341/175341 [==============================] - 310s 2ms/step - loss: 1.3135 - accuracy: 0.5510 - val_loss: 1.2747 - val_accuracy: 0.5922
Epoch 6/200
175341/175341 [==============================] - 291s 2ms/step - loss: 1.3067 - accuracy: 0.5521 - val_loss: 1.2738 - val_accuracy: 0.5804
Epoch 7/200
175341/175341 [================

Epoch 108/200
175341/175341 [==============================] - 254s 1ms/step - loss: 0.7885 - accuracy: 0.6762 - val_loss: 0.8332 - val_accuracy: 0.6151
Epoch 109/200
175341/175341 [==============================] - 251s 1ms/step - loss: 0.7770 - accuracy: 0.6844 - val_loss: 0.8117 - val_accuracy: 0.6309
Epoch 110/200
175341/175341 [==============================] - 255s 1ms/step - loss: 0.7646 - accuracy: 0.6926 - val_loss: 0.7912 - val_accuracy: 0.6584
Epoch 111/200
175341/175341 [==============================] - 254s 1ms/step - loss: 0.7453 - accuracy: 0.7016 - val_loss: 0.7769 - val_accuracy: 0.6480
Epoch 112/200
175341/175341 [==============================] - 254s 1ms/step - loss: 0.7441 - accuracy: 0.7024 - val_loss: 0.8141 - val_accuracy: 0.6969
Epoch 113/200
175341/175341 [==============================] - 251s 1ms/step - loss: 0.7343 - accuracy: 0.7081 - val_loss: 0.7716 - val_accuracy: 0.6489
Epoch 114/200
175341/175341 [==============================] - 253s 1ms/step - los

In [16]:
from sklearn.metrics import accuracy_score
predictions = cnn.predict([xdelta1])

In [31]:
import sys
np.set_printoptions(threshold=sys.maxsize)
pred = np.zeros((82332,1), dtype=int)
for i in range(0, predictions.shape[0]):
    pred[i] = np.where(predictions[i] == np.amax(predictions[i]))

In [33]:
pred = to_categorical(pred)
z = np.zeros((82332,1), dtype=int)
pred = np.append(pred, z, axis=1)

In [34]:
pred.shape


(82332, 10)

In [35]:
print("Accuracy: ",accuracy_score(y,pred))

Accuracy:  0.7007967740368265


In [36]:
from sklearn.metrics import multilabel_confusion_matrix,classification_report,confusion_matrix
print("Classification Report: \n",classification_report(y,pred))
conmat = np.array(confusion_matrix(y.argmax(axis=1), pred.argmax(axis=1)))
confusion = pd.DataFrame(conmat, index=['0','1','2','3','4','5','6','7','8','9'],
                        columns=[['0','1','2','3','4','5','6','7','8','9']])
print (conmat)
conmat.shape

Classification Report: 
               precision    recall  f1-score   support

           0       0.11      0.00      0.00       677
           1       1.00      0.01      0.01       583
           2       0.61      0.02      0.04      4089
           3       0.44      0.82      0.58     11132
           4       0.25      0.51      0.33      6062
           5       1.00      0.96      0.98     18871
           6       0.92      0.67      0.77     37000
           7       0.67      0.71      0.69      3496
           8       0.22      0.12      0.16       378
           9       0.00      0.00      0.00        44

   micro avg       0.70      0.70      0.70     82332
   macro avg       0.52      0.38      0.36     82332
weighted avg       0.79      0.70      0.71     82332
 samples avg       0.70      0.70      0.70     82332

[[    1     0     0   658     2     0    16     0     0     0]
 [    0     4     0   544    18     0    15     2     0     0]
 [    0     0    75  3460   308     

/home/ritika/.virtualenvs/project/lib/python3.5/site-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


(10, 10)

Classification Report: 
               precision    recall  f1-score   support

           0       0.11      0.00      0.00       677
           1       1.00      0.01      0.01       583
           2       0.61      0.02      0.04      4089
           3       0.44      0.82      0.58     11132
           4       0.25      0.51      0.33      6062
           5       1.00      0.96      0.98     18871
           6       0.92      0.67      0.77     37000
           7       0.67      0.71      0.69      3496
           8       0.22      0.12      0.16       378
           9       0.00      0.00      0.00        44

   micro avg       0.70      0.70      0.70     82332
   macro avg       0.52      0.38      0.36     82332
weighted avg       0.79      0.70      0.71     82332
 samples avg       0.70      0.70      0.70     82332

[[    1     0     0   658     2     0    16     0     0     0]
 [    0     4     0   544    18     0    15     2     0     0]
 [    0     0    75  3460   308     6   215    22     3     0]
 [    0     0    17  9114   906     0  1019    63    13     0]
 [    0     0     4  1914  3104     0   753   256    31     0]
 [    0     0    12   403   149 18146   148    10     3     0]
 [    8     0     6  3574  7714     1 24735   851   111     0]
 [    0     0     6   726   253     4    34  2473     0     0]
 [    0     0     2    91   187     6    20    26    46     0]
 [    0     0     0    34     4     0     6     0     0     0]]
